Mathlib/LinearAlgebra/Eigenspace/Matrix.lean

```diff
@@ -69,6 +69,50 @@ lemma hasEigenvalue_toLin'_diagonal_iff [NoZeroDivisors R] (d : n → R) {μ : R
 
 end NontrivialCommRing
 
+namespace Matrix
+
+variable [CommRing R] [IsDomain R] {d : n → R}
+
+@[simp]
+lemma maxGenEigenspace_diagonal_eq_bot_iff {μ : R} :
+    maxGenEigenspace (diagonal d).toLin' μ = ⊥ ↔ μ ∉ Set.range d := by
+  rw [← not_iff_not]
+  simp only [Set.mem_range, not_exists, not_forall, not_not, ← ne_eq, Submodule.ne_bot_iff]
+  refine ⟨fun ⟨x, hx, hx₀⟩ ↦ ?_, ?_⟩
+  · obtain ⟨i, hi⟩ : ∃ i, x i ≠ 0 := Function.ne_iff.mp hx₀
+    use i
+    have : (diagonal d).toLin' - μ • 1 = (diagonal (d - μ • 1)).toLin' := by
+      aesop (add simp Pi.single_apply)
+    obtain ⟨k, hk⟩ : ∃ k, diagonal ((d - μ • 1) ^ k) *ᵥ x = 0 := by
+      simpa only [mem_maxGenEigenspace, this, ← toLin'_pow, toLin'_apply, diagonal_pow] using hx
+    replace hk : x i * (d i - μ) ^ k = 0 := by
+      simpa [mulVec_eq_sum, diagonal_apply] using congr_fun hk i
+    simp only [mul_eq_zero, hi, pow_eq_zero_iff', ne_eq, false_or, sub_eq_zero] at hk
+    exact hk.1
+  · rintro ⟨i, rfl⟩
+    refine ⟨Pi.single i 1, (mem_maxGenEigenspace _ _ _).mpr ⟨1, ?_⟩, by simp⟩
+    ext j
+    simp [Pi.single_apply]
+
+@[simp]
+lemma iSup_maxGenEigenspace_diagonal_eq_top :
+    ⨆ μ, maxGenEigenspace (diagonal d).toLin' μ = ⊤ := by
```

Comment:
Likewise, you can prove that for `i : n`, `eigenspace (diagonal d) μ = span (Pi.basisFun R n '' (d ⁻¹' μ))`. Or else you can just use the result in the other comment and [hasEigenvector_toLin'_diagonal](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/Eigenspace/Matrix.html#hasEigenvector_toLin'_diagonal) to prove that every element of `Pi.basisFun` is in the LHS.

Reply:
As above, I've incorporated this suggestion.